Common/Python/PandasConverter.cs

```diff
@@ -115,31 +116,46 @@ public PyObject GetDataFrame<T>(IEnumerable<T> data)
                 var index = CreateIndex(data.First().Symbol, data.Select(x => x.Time));
 
                 var pyDict = new PyDict();
-                
-                pyDict.SetItem("low", _pandas.Series(data.Select(x => (double)x.Low).ToList(), index));
-                pyDict.SetItem("open", _pandas.Series(data.Select(x => (double)x.Open).ToList(), index));
-                pyDict.SetItem("high", _pandas.Series(data.Select(x => (double)x.High).ToList(), index));
-                pyDict.SetItem("close", _pandas.Series(data.Select(x => (double)x.Close).ToList(), index));
 
                 if (typeof(T) == typeof(TradeBar))
                 {
-                    Func<IBaseDataBar, double> getVolume = x => { var bar = x as TradeBar; return (double)bar.Volume; };
-                    pyDict.SetItem("volume", _pandas.Series(data.Select(x => getVolume(x)).ToList(), index));
+                    var bars = data.Select(x => x as TradeBar);
+                    pyDict.SetItem("low", _pandas.Series(bars.Select(x => (double)x.Low).ToList(), index));
+                    pyDict.SetItem("open", _pandas.Series(bars.Select(x => (double)x.Open).ToList(), index));
+                    pyDict.SetItem("high", _pandas.Series(bars.Select(x => (double)x.High).ToList(), index));
+                    pyDict.SetItem("close", _pandas.Series(bars.Select(x => (double)x.Close).ToList(), index));
+                    pyDict.SetItem("volume", _pandas.Series(bars.Select(x => (double)x.Volume).ToList(), index));
```

Comment:
>NOTE: This is an existing issue with the `PandasConverter` that perhaps should be looked at separately.

Each one of these `bars.Select` statement causes an extra enumeration of the `bars` enumerable. To make matters worse, this is all invoked 4*N times, where N is the number of data points.

I'm wondering if we need to invoke `GetDataFrame<T>` multiple times, or perhaps we just invoke it once and check the type of each item.

Ideally this would be implemented in such a way that we only cause one iteration of the enumerable. During this single iteration we would be building up all the various lists required to instantiate the pandas data frame.

Reply:
Agree. We can talk how would be the best way to do it.
I have made an experiment, where I would rewrite the `IEnumerable<Slice>` as a `Dictionary<string,List<double>>` that could be used to set the items in the PyDict object, but it got slower (I didn't make ver extensive performance tests).

Reply:
awesome that you benchmarked it! What kind of throughput is `GetDataFrame(IEnumerable<Slice> slice)` method getting?  I'd be interested in timings containing 100 data points for 1, 10 and 100 symbols. I suspect we have an exponential slowdown here (w.r.t num data points and/or num symbols)

I'll take a peek at this later -- I'm fairly convinced we can get it to a single iteration. One idea is this method could accept `IEnumerable` instead of `IEnumerable<T>` and while it enumerates (`while(enumerator.MoveNext())`) it can check each type `var type = enumerator.Current.GetType()` -- there's lots of options here.

Reply:
That method is returning a `pandas.DataFrame` (it is represented as a `PyObject` object). That dataframe is an enhanced dictionary that is created based on a python dictionary (`PyDict` object).